### PR TITLE
Eliminate `services.Services` interface and `XXXClient()` methods on `auth.Services`

### PR DESCRIPTION
**Pull request by strideynet**
I was recently adding a new resource and noticed how it was a little awkward adding these `XXXClient()` methods onto `auth.Services`, upon further probing, it looked like these were only necessary for configuring the cache. This PR feels to me like an improvement, but I'm mostly putting this out there for feedback since I know there might be some varied opinions on this.

Pros:

- Eliminates a shared interface defined at producer side (`services.Services`), that didn't seem to be implemented naturally by `auth.Services` or `authclient.ClientI` (both required some kind of adapting)
- Removes `XXXClient()` helper methods from `auth.Services`
  - These seemed to only exist to make `auth.Services` implement `services.Services`.
  - However, because they exist, they add another way of accessing a service. Several places within the `auth` package would do something like `a.FoosClient().GetFoo()` rather than `a.Foos.GetFoo()`. Sometimes these `Client()` methods would return nil (e.g `SecReportsClient()`, which seems like a foot gun. We already have too many unclear ways of accessing services ( due to their embedding ), and this only made matters worse.
- Removes `accessPointWrapper` which seemed to only exist to make `authclient.ClientI` implement `services.Services`.
- Cache setup for auth.Services now more explicitly references the individual service rather than relying on embedding.

Cons:

- Set up of cache config is now "duplicated" for authclient.ClientI and authservices.Services.

### CODE REVIEW

**Comment by strideynet** [comment]
If this looks good, I'll also raise an `e` PR to replace the single callsite to `auth.Services.OktaClient()` so that helper method can also be removed.

File: lib/auth/auth.go [hunk 1]
```diff
@@ -621,9 +621,48 @@ type Services struct {
 	services.DevicesGetter
 }
 
-// SecReportsClient returns the security reports client.
-func (r *Services) SecReportsClient() *secreport.Client {
-	return nil
+// NewAccessCacheForServices creates a new cache for a Teleport service that
+// uses the provided services to populate its resource collections.
+func NewAccessCacheForServices(cfg accesspoint.AccessCacheConfig, services *Services) (*cache.Cache, error) {
```
Comment (strideynet) [inline lib/auth/auth.go:626]: I'd originally wanted this function to be in `accesspoint`, but unfortunately, this creates an import loop due to `NewTestAuthServer` in `helpers.go`.

File: lib/auth/accesspoint/accesspoint.go [hunk 2]
```diff
@@ -83,9 +77,54 @@ func (c *AccessCacheConfig) CheckAndSetDefaults() error {
 	return nil
 }
 
-// NewAccessCache builds a cache.Cache instance for a teleport service. This logic has been
-// broken out of lib/service in order to support easier unit testing of process components.
-func NewAccessCache(cfg AccessCacheConfig) (*cache.Cache, error) {
+// NewAccessCacheForClient creates a new cache for a teleport service that
+// uses an authclient.ClientI to populate its resource collections.
+func NewAccessCacheForClient(cfg AccessCacheConfig, client authclient.ClientI) (*cache.Cache, error) {
```
Comment (tigrato) [inline lib/auth/accesspoint/accesspoint.go:82]: Can we finally rename this function to `NewCacheForClient`?

File: lib/auth/auth.go (hunk 1 above)
Comment (tigrato) [reply lib/auth/auth.go:626]: Refactoring `auth.Services` out of auth package would solve the problem and simplify the `auth` package itself

Comment (strideynet) [reply lib/auth/auth.go]: [ad203d8](https://github.com/gravitational/teleport/pull/45058/commits/ad203d889881203b6eddc6e5c4f97fb13a5bf268)

File: lib/auth/accesspoint/accesspoint.go (hunk 2 above)
Comment (strideynet) [reply lib/auth/accesspoint/accesspoint.go]: 5519c635a1

File: lib/auth/auth.go [hunk 3]
```diff
@@ -578,128 +578,7 @@ func NewServer(cfg *InitConfig, opts ...ServerOption) (*Server, error) {
...
-// GetAccessGraphSecretsGetter returns the AccessGraph secrets service.
-func (r *Services) GetAccessGraphSecretsGetter() services.AccessGraphSecretsGetter {
-	return r.AccessGraphSecretsGetter
-}
-
-// GetDevicesGetter returns the trusted devices service.
-func (r *Services) GetDevicesGetter() services.DevicesGetter {
-	return r.DevicesGetter
-}
+type Services = authservices.Services
```
Comment (strideynet) [inline lib/auth/auth.go:581]: I'll do another PR which actually changes the references - for now - this prevents breaking the teleport-e build and avoids this PR having changes across hundreds of files.

File: lib/auth/accesspoint/accesspoint.go [hunk 4]
```diff
@@ -83,9 +78,97 @@ func (c *AccessCacheConfig) CheckAndSetDefaults() error {
 	return nil
 }
 
-// NewAccessCache builds a cache.Cache instance for a teleport service. This logic has been
-// broken out of lib/service in order to support easier unit testing of process components.
-func NewAccessCache(cfg AccessCacheConfig) (*cache.Cache, error) {
+// NewCacheForServices creates a new cache for a Teleport service that
+// uses the provided services to populate its resource collections.
+func NewCacheForServices(
```
Comment (codingllama) [inline lib/auth/accesspoint/accesspoint.go:83]: Should the methods here be NewCache<i>From</i>Services / NewCache<i>From</i>Client ?

File: lib/auth/accesspoint/accesspoint.go [hunk 5]
```diff
@@ -83,9 +78,97 @@ func (c *AccessCacheConfig) CheckAndSetDefaults() error {
...
+	cfg Config, services *authservices.Services,
```
Comment (codingllama) [inline lib/auth/accesspoint/accesspoint.go:84]: Suggestion: Pass cfg last, so invocations that pass a literal Config{} list other parameters first?

**Review by codingllama — COMMENTED** [review]
Seems like an improvement to me. +1.

File: lib/auth/accesspoint/accesspoint.go (hunk 4 above)
Comment (strideynet) [reply lib/auth/accesspoint/accesspoint.go:83]: Works for me.

File: lib/auth/services/services.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+	return r.Identity.WebTokens().Get(ctx, req)
+}
+
+// GenerateAWSOIDCToken generates a token to be used to execute an AWS OIDC Integration action.
+func (r *Services) GenerateAWSOIDCToken(ctx context.Context, integration string) (string, error) {
+	return r.IntegrationsTokenGenerator.GenerateAWSOIDCToken(ctx, integration)
+}
+
+// OktaClient returns the okta client.
+// TODO(noah): Used in teleport-e - could be removed once e is updated to just
```
Comment (strideynet) [inline lib/auth/services/services.go:78]: https://github.com/gravitational/teleport.e/pull/4838

File: lib/auth/accesspoint/accesspoint.go (hunk 5 above)
Comment (strideynet) [reply lib/auth/accesspoint/accesspoint.go]: [340b808](https://github.com/gravitational/teleport/pull/45058/commits/340b808630be3b04c8a8190d5c0320ce7003c23a)

File: lib/auth/accesspoint/accesspoint.go (hunk 4 above)
Comment (strideynet) [reply lib/auth/accesspoint/accesspoint.go]: [340b808](https://github.com/gravitational/teleport/pull/45058/commits/340b808630be3b04c8a8190d5c0320ce7003c23a)

**Review by codingllama — APPROVED** [review]

**Review by tigrato — APPROVED** [review]

**Comment by strideynet** [comment]
@rosstimothy I'll hold off merging to give you a chance to review.

**Review by rosstimothy — APPROVED** [review]

File: lib/auth/accesspoint/accesspoint.go [hunk 7]
```diff
@@ -83,9 +78,95 @@ func (c *AccessCacheConfig) CheckAndSetDefaults() error {
...
+	cacheCfg.Okta = services.Okta
+	cacheCfg.AccessLists = services.AccessLists
+	cacheCfg.AccessMonitoringRules = services.AccessMonitoringRules
+	cacheCfg.SecReports = services.SecReports
+	cacheCfg.UserLoginStates = services.UserLoginStates
+	cacheCfg.Integrations = services.Integrations
+	cacheCfg.DiscoveryConfigs = services.DiscoveryConfigs
+	cacheCfg.WebSession = services.Identity.WebSessions()
+	cacheCfg.WebToken = services.Identity.WebTokens()
+	cacheCfg.KubeWaitingContainers = services.KubeWaitingContainer
```
[the remark this answers is not in the document]
Comment (strideynet) [reply lib/auth/accesspoint/accesspoint.go]: [dc7d3ef](https://github.com/gravitational/teleport/pull/45058/commits/dc7d3ef9f7965ac7d4d74c1b1930752257f15c5c)

**Pull request merged**